### PR TITLE
feat(ptv3)!: export precomputed serialized-pooling metadata for ONNX

### PR DESCRIPTION
**Pull request by mojomex**
## Summary

> [!WARNING]
> This breaks backward-compatibility with current PTv3 ONNX files.

Adapts the PTv3 ONNX export to match the updated inference node in
[autowarefoundation/autoware_universe#12727](https://github.com/autowarefoundation/autoware_universe/pull/12727)
("precompute serialized pooling metadata").

PTv3's `SerializedPooling`, when exported 1:1 from Pointcept, discovers the
unique parent voxels in-graph with `Unique`, whose output count is
data-dependent. TensorRT cannot infer those shapes ahead of time, so it inserts
`trainStation`s (CPU/GPU syncs + allocations) that hurt latency.

This PR moves that discovery to Autoware preprocessing. Each encoder pooling
stage now receives its pooling metadata as additional ONNX inputs, so the
exported pooling subgraph is just native `Gather` + the existing
`autoware::SegmentCSR` plugin — no `Unique`, no data-dependent shapes.

## Changes

- **`tools/export.py`** — compute per-stage serialized-pooling metadata from the
  sample frame and register the tensors as named, dynamic ONNX inputs
  (`serialized_pooling_{i}_*`). The tensors are `forward()` arguments, so tracing
  records them as graph inputs.
- **`models/.../point_transformer_v3m1_base.py`** — in `export_mode`,
  `SerializedPooling` reads its metadata from the `Point` instead of computing
  `Unique`/`argsort`/`segment_csr`. The metadata travels as a
  `SerializedPoolingMeta` **dataclass** (not a dict) and is propagated across
  stages: addict's `_hook` only auto-converts plain `dict` values into `Point`
  objects, and a coord-less metadata dict would recurse infinitely
  (`Point.__init__` requires `"coord"`), whereas a dataclass is stored verbatim.
- **`tests/test_serialized_pooling.py`** — CPU test asserting the `export_mode`
  output matches the training-time path (`feat`, `grid_coord`,
  `serialized_order/inverse`, `batch`, `sparse_shape`, `pooling_inverse`).
- **`README.md`** — documents the ONNX preprocessing contract.

## ONNX / Engine interface

For each pooling stage `i` (input voxels `N_i`, pooled voxels `M_i`,
serialization orders `O`):

| Input | Type | Shape | Meaning |
|---|---|---:|---|
| `serialized_pooling_i_indices` | int64 | `[N_i]` | `Gather` indices into CSR segment order |
| `serialized_pooling_i_indptr` | int64 | `[M_i + 1]` | `autoware::SegmentCSR` row pointer |
| `serialized_pooling_i_cluster` | int64 | `[N_i]` | input voxel → pooled voxel id (unpooling) |
| `serialized_pooling_i_head_indices` | int64 | `[M_i]` | representative input voxel per pooled voxel |
| `serialized_pooling_i_grid_coord` | int64 | `[M_i, 3]` | pooled voxel coordinates |
| `serialized_pooling_i_serialized_order` | int64 | `[O, M_i]` | serialization order over pooled voxels |
| `serialized_pooling_i_serialized_inverse` | int64 | `[O, M_i]` | inverse serialization order |

Original inputs `grid_coord [N_0,3]`, `feat [N_0,4]`, `serialized_code [O,N_0]`
remain. The exported model uses `O = 2` (`["z", "z-trans"]`) and 4 stride-2
pooling stages.

## Validation

Exported a j6gen2 lidarseg checkpoint in the `awml-ptv3` docker:

```bash
python projects/PTv3/tools/export.py \
  --config-file projects/PTv3/configs/semseg-pt-v3m1-0-t4dataset.py --num-gpus 1 \
  --options save_path=work_dirs/experiment weight=work_dirs/ptv3/model/model_best.pth
```

`onnx.checker` of the result passes, and the graph matches the contract above:

- 31 graph inputs = `grid_coord`/`feat`/`serialized_code` + 4 pooling stages × 7
  metadata inputs — all present, correct dtypes/shapes.
- 4× `autoware::SegmentCSR` + native `Gather` for pooling.
- **no `Unique`** node. (The 2 remaining `Argsort` ops are the full-resolution
  initial serialization — static shape, unchanged from the previous export.)
- unit test `test_serialized_pooling.py` passes.

## Notes

- `serialized_depth` is no longer a graph input (it was only used for
  constant-folded scalar math).
- `serialized_pooling_i_indptr` (length `M_i + 1`) currently shares the
  `..._out_voxels` symbolic dim name with the `M_i`-length inputs. This is
  cosmetic — the Autoware node binds explicit per-tensor shapes — but can be
  given a distinct name if preferred.

### CODE REVIEW

File: projects/PTv3/models/point_transformer_v3/point_transformer_v3m1_base.py [hunk 1]
```diff
@@ -403,6 +404,25 @@ def forward(self, point: Point):
...
 
 
+@dataclass
+class SerializedPoolingMeta:
+    """Per-stage serialized-pooling metadata supplied as ONNX graph inputs during export.
+
+    Stored as a dataclass (not a dict) so it can travel inside the addict-based ``Point``:
+    addict only auto-converts plain ``dict`` values into ``Point`` objects, which would recurse
+    infinitely here because the metadata has no ``coord`` key. A dataclass instance is left
+    untouched by addict's ``_hook``, so it survives ``Point`` construction and reconstruction.
```
Comment (mojomex) [inline projects/PTv3/models/point_transformer_v3/point_transformer_v3m1_base.py]: remove lengthy explanation.

File: projects/PTv3/tests/test_serialized_pooling.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                raise NotImplementedError(f"Unsupported reduce mode: {reduce}")
+        return torch.stack(outputs, dim=0)
+
+    torch_scatter.segment_csr = _segment_csr
+    sys.modules["torch_scatter"] = torch_scatter
+
+
+PROJECT_ROOT = Path(__file__).resolve().parents[1]
+if str(PROJECT_ROOT) not in sys.path:
+    sys.path.insert(0, str(PROJECT_ROOT))
```
Comment (mojomex) [inline projects/PTv3/tests/test_serialized_pooling.py]: Remove hacky fallbacks and discovery. Use the docker container to get the correct environment.

File: projects/PTv3/tests/test_serialized_pooling.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    grid_coord = point.grid_coord.index_select(0, head_indices) >> depth
+    serialized_code = pooled_code.index_select(1, head_indices)
+    serialized_order = torch.stack([argsort(code) for code in serialized_code], dim=0)
+    serialized_inverse = torch.zeros_like(serialized_order).scatter_(
+        dim=1,
+        index=serialized_order,
+        src=torch.arange(0, serialized_code.shape[1], device=serialized_order.device).repeat(
+            serialized_code.shape[0], 1
+        ),
+    )
```
Comment (mojomex) [inline projects/PTv3/tests/test_serialized_pooling.py]: hard to understand in this long shape and without comments. Refactor.

File: projects/PTv3/tools/export.py [hunk 4]
```diff
@@ -7,13 +9,114 @@
...
 from engines.train import TRAINERS
+from models.point_transformer_v3.point_transformer_v3m1_base import SerializedPoolingMeta
 from models.scatter.functional import argsort
 from models.utils.structure import Point, bit_length_tensor
 from torch.nn import functional as F
 
 # NOTE: keep this import last; it overrides sparse conv registration for export.
 import SparseConvolution  # isort: skip
 
+# Per-stage ONNX input field order, kept in sync with the model-side dataclass.
```
Comment (mojomex) [inline projects/PTv3/tools/export.py]: remove chatty comment

File: projects/PTv3/tests/test_serialized_pooling.py (hunk 3 above)
Comment (mojomex) [reply projects/PTv3/tests/test_serialized_pooling.py]: Also, re-implements exactly what exporter does. De-duplicate.

File: projects/PTv3/tools/export.py [hunk 5]
```diff
@@ -7,13 +9,114 @@
...
+    next_serialized_order = torch.stack([argsort(code) for code in next_serialized_code], dim=0)
+    next_serialized_inverse = torch.zeros_like(next_serialized_order).scatter_(
+        dim=1,
+        index=next_serialized_order,
+        src=torch.arange(
+            0,
+            next_serialized_code.shape[1],
+            device=next_serialized_order.device,
+        ).repeat(next_serialized_code.shape[0], 1),
+    )
```
Comment (mojomex) [inline projects/PTv3/tools/export.py]: Spaghetti-like code without comments. refactor.

File: projects/PTv3/tools/export.py [hunk 6]
```diff
@@ -7,13 +9,114 @@
...
+        "serialized_order": next_serialized_order,
+        "serialized_inverse": next_serialized_inverse,
+    }
+
+
+def build_serialized_pooling_inputs(grid_coord, serialized_code, serialized_order, strides):
+    """Build sample tensors for the ONNX/inference serialized-pooling contract.
+
+    These tensors are not constants in the exported graph. They are sample inputs used by
+    torch.onnx.export to define the interface that Autoware preprocessing fills at runtime.
```
Comment (mojomex) [inline projects/PTv3/tools/export.py]: remove `Autoware`

File: projects/PTv3/tools/export.py [hunk 7]
```diff
@@ -7,13 +9,114 @@
...
+        for field in SERIALIZED_POOLING_FIELDS:
+            name = prefix + field
+            flat_inputs.append(metadata[field])
+            input_names.append(name)
+            if field in {"grid_coord", "serialized_order", "serialized_inverse"}:
+                axis = 0 if field == "grid_coord" else 1
+                dynamic_axes[name] = {axis: f"serialized_pooling_{stage_index}_out_voxels"}
+            elif field in {"head_indices", "indptr"}:
+                dynamic_axes[name] = {0: f"serialized_pooling_{stage_index}_out_voxels"}
+            else:
```
Comment (mojomex) [inline projects/PTv3/tools/export.py]: Enumerate fields exhaustively and have an `else` that catches any unexpected fields, throwing an assertionError.

File: projects/PTv3/tools/export.py [hunk 8]
```diff
@@ -58,6 +171,9 @@ def forward(
             "serialized_code": serialized_code,
             "serialized_order": serialized_order,
             "serialized_inverse": serialized_inverse,
+            # List of SerializedPoolingMeta dataclasses; addict stores these verbatim (only plain
+            # dicts get auto-converted to Point, which would recurse on the missing "coord").
```
Comment (mojomex) [inline projects/PTv3/tools/export.py]: remove comment

File: projects/PTv3/models/point_transformer_v3/point_transformer_v3m1_base.py (hunk 1 above)
Comment (mojomex) [reply projects/PTv3/models/point_transformer_v3/point_transformer_v3m1_base.py]: Shortened to a one-line docstring in cbd4a1cf.

File: projects/PTv3/tests/test_serialized_pooling.py (hunk 2 above)
Comment (mojomex) [reply projects/PTv3/tests/test_serialized_pooling.py]: Dropped the `inference_mode` / `torch_scatter` / `sys.path` shims (and the `assert_close` fallback); the test now relies on the docker environment. cbd4a1cf

File: projects/PTv3/tests/test_serialized_pooling.py (hunk 3 above)
Comment (mojomex) [reply projects/PTv3/tests/test_serialized_pooling.py]: De-duplicated: the test now reuses the shared `build_serialized_pooling_meta` from the model module instead of reimplementing the metadata math. cbd4a1cf

File: projects/PTv3/tools/export.py (hunk 4 above)
Comment (mojomex) [reply projects/PTv3/tools/export.py]: Removed in cbd4a1cf.

File: projects/PTv3/tools/export.py (hunk 5 above)
Comment (mojomex) [reply projects/PTv3/tools/export.py]: Refactored into clear, commented steps and extracted the shared `build_serialized_pooling_meta` into the model module. cbd4a1cf

File: projects/PTv3/tools/export.py (hunk 6 above)
Comment (mojomex) [reply projects/PTv3/tools/export.py]: Removed in cbd4a1cf.

File: projects/PTv3/tools/export.py (hunk 7 above)
Comment (mojomex) [reply projects/PTv3/tools/export.py]: Done — `_serialized_pooling_dynamic_axis` now enumerates every field and raises `AssertionError` on unexpected ones. cbd4a1cf

File: projects/PTv3/tools/export.py (hunk 8 above)
Comment (mojomex) [reply projects/PTv3/tools/export.py]: Removed in cbd4a1cf.

File: projects/PTv3/tools/export.py [hunk 9]
```diff
@@ -7,13 +9,54 @@
...
+def _serialized_pooling_dynamic_axis(stage_index, field):
+    """Return the single dynamic ONNX axis for one metadata field of one pooling stage."""
+    in_voxels = f"serialized_pooling_{stage_index}_in_voxels"
+    out_voxels = f"serialized_pooling_{stage_index}_out_voxels"
+    if field in ("indices", "cluster"):
+        return {0: in_voxels}
+    if field in ("indptr", "head_indices", "grid_coord"):
+        return {0: out_voxels}
+    if field in ("serialized_order", "serialized_inverse"):
+        return {1: out_voxels}
```
[the remark this answers is not in the document]
Comment (mojomex) [reply projects/PTv3/tools/export.py]: Fixed in b04cf867 — `serialized_pooling_i_indptr` now gets its own symbolic dim `serialized_pooling_i_out_voxels_plus_one` (it holds M+1 entries), so it no longer shares the M-length `out_voxels` dim.

File: projects/PTv3/models/point_transformer_v3/point_transformer_v3m1_base.py [hunk 10]
```diff
@@ -450,39 +504,46 @@ def forward(self, point: Point):
...
             _, indices = torch.sort(cluster)
+            idx_ptr = torch.cat([counts.new_zeros(1), torch.cumsum(counts, dim=0)])
         else:
-            code_, cluster, counts, num_unique = unique(code[0])
-            indices = argsort(cluster)
+            assert self.export_stage_index is not None
+            metadata = point.serialized_pooling[self.export_stage_index]
+            indices = metadata.indices
+            idx_ptr = metadata.indptr
+            cluster = metadata.cluster
```
[the remark this answers is not in the document]
Comment (mojomex) [reply projects/PTv3/models/point_transformer_v3/point_transformer_v3m1_base.py:526]: Added in b04cf867 — export mode now asserts that `serialized_pooling` is present and covers the current stage, so a missing or too-short metadata list fails with an actionable message.

File: projects/PTv3/README.md [hunk 11]
```diff
@@ -63,6 +63,26 @@ python projects/PTv3/tools/export.py --config-file projects/PTv3/configs/semseg-
 
 which will generate a file called `ptv3.onnx`
 
+### ONNX preprocessing contract
+
+The exported ONNX expects PTv3 serialized-pooling metadata to be generated outside the engine.
+For every encoder pooling stage `i`, preprocessing must provide:
+
```
[the remark this answers is not in the document]
Comment (mojomex) [reply projects/PTv3/README.md]: Added the base inputs to the README in b04cf867 (`grid_coord`, `feat`, `serialized_code`). One correction: `serialized_depth` is not a runtime input — it is constant-folded out during export, so the exported graph only takes `grid_coord`/`feat`/`serialized_code` as base inputs (matching autoware_universe#12727). The README now states this explicitly. Leaving this thread open for review.

**Comment by mojomex** [comment]
@amadeuszsz Since 
* tier4/autoware-ml#46
has been merged, could you approve/take a look at this one as well? For compatibility until everyone switches over.

**Review by amadeuszsz — APPROVED** [review]
LGTM
